### PR TITLE
Added commands to switch storage provider

### PR DESCRIPTION
**Pull request by bastien70**
See #51 

- Command to manually switch to local storage mode
- Command to manually switch to AWS S3 storage mode
- Storage type initialization at installation

### CODE REVIEW

**Comment by jmsche** [comment]
Why not a single command instead of two?

Also I would not put it in the post install command.

File: src/Command/SwitchStorageModeCommand.php [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    description: 'Switch storage mode',
+)]
+final class SwitchStorageModeCommand extends AbstractDotEnvCommand
+{
+    public const FILE_DIR = 'config/packages';
+    public const FILE_NAME = 'vich_uploader';
+
+    public const STORAGE_LOCALLY = 'Locally';
+    public const STORAGE_AWS = 'AWS S3';
+    private YamlService $yamlService;
```
Comment (jmsche) [inline src/Command/SwitchStorageModeCommand.php:28]:
```suggestion
    public const STORAGE_AWS = 'AWS S3';

    private YamlService $yamlService;
```

File: src/Command/SwitchStorageModeCommand.php [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+                ],
+            ],
+        ];
+
+        $this->yamlService->writeFileContent($content);
+    }
+
+    private function updateEnvVariables(string $key, string $question, ?string $default = null): void
+    {
+        $editor = $this->editor;
```
Comment (jmsche) [inline src/Command/SwitchStorageModeCommand.php:113]: Is this really needed?

File: src/Service/YamlService.php [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+namespace App\Service;
+
+use Symfony\Component\Finder\Finder;
+use Symfony\Component\Yaml\Yaml;
+
+class YamlService
+{
+    public const CONFIG_INLINE_LEVEL = 5;
+    private ?string $absoluteFilePath;
```
Comment (jmsche) [inline src/Service/YamlService.php:13]:
```suggestion
    public const CONFIG_INLINE_LEVEL = 5;

    private ?string $absoluteFilePath;
```

File: src/Service/YamlService.php [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+    private Finder $finder;
+
+    public function __construct(string $fileDir, string $fileName)
+    {
+        $this->finder = new Finder();
+        $this->setAbsoluteFilePath($this->buildFilePath($fileDir, $fileName));
```
Comment (jmsche) [inline src/Service/YamlService.php:20]: Not sure why you're using the Finder instead of providing the direct path?

**Review by jmsche — CHANGES_REQUESTED** [review]

**Comment by jmsche** [comment]
Hi,

I'm not sure this PR (or even how it is currently documented) is the best way to handle an issue: files that are committed to this repository should not be modified.

And, with this PR or the way it's currently documented, project files are modified, implying that the final user can't rebase/upgrade easily.

Maybe one way to fix it correctly would be writing a compiler pass instead?

**Comment by bastien70** [comment]
> Hi,
> 
> I'm not sure this PR (or even how it is currently documented) is the best way to handle an issue: files that are committed to this repository should not be modified.
> 
> And, with this PR or the way it's currently documented, project files are modified, implying that the final user can't rebase/upgrade easily.
> 
> Maybe one way to fix it correctly would be writing a compiler pass instead?

Hello,

Yes you are right, during development, inevitably, the configuration file was modified and committed.

But it should be the only feature capable of modifying the configuration file, so afterwards there should be no problems with it. Therefore, the idea could be, just before validating this PR, to put the configuration file back with the default local storage mode.

I've never used compiler pass, are you sure it has to come to this? (given what I said above, since the file shouldn't be changed in future PRs)

**Comment by jmsche** [comment]
> But it should be the only feature capable of modifying the configuration file, so afterwards there should be no problems with it. Therefore, the idea could be, just before validating this PR, to put the configuration file back with the default local storage mode.

The problem is the file will be modified by the end user, hence blocking their next git pull.

> I've never used compiler pass, are you sure it has to come to this? (given what I said above, since the file shouldn't be changed in future PRs)

Well, I don't see an other way to change the configuration of a bundle without changing a YAML file?

**Comment by bastien70** [comment]
Okay I just read the documentation on the compiler pass. So if I understand correctly, it allows us to modify certain configuration values, services, etc.

So we would run it from the kernel?

I have never used this kind of thing. If you know exactly how to do it, could you take over on this PR?

**Comment by jmsche** [comment]
I'll see what I can do this week

**Comment by bastien70** [comment]
I close this PR because the feature will no longer be an idea.

**Pull request closed without merging**